### DISCUSSION TITLE
Enforcing linear history

### DISCUSSION THREAD

**Discussion by DavidBiddle** [govuk-forms/forms#39] [Polls]
Should we turn on enforcing [linear history](https://docs.github.com/en/github-ae@latest/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/about-protected-branches#require-linear-history) in Github?

This would mean that we could only use the 'Squash and Merge' or 'Rebase and merge' pull request merge strategies. The advantage of this is that it can be easier to revert changes done this way than it is to revert changes done via merge commit. 

[This discussion originally contained some information from the GDS Way which I thought backed up this proposal, it later transpired that I had misinterpreted this]

I personally find that linear history makes it easier to see at a glance what changes have been made, and it's easier to use tools like `git bisect`.

**Comment by DavidBiddle** [comment]
We discussed this in a tech catch up in February and clarified that we should be:
- using rebases / history rewriting etc in our local feature branches to keep history tidy (i.e. rebase onto `main`, don't merge `main` into your branch)
- using merge commits to incorporate our feature branches into `main`, rather than rebasing or squashing into `main`.

This is in keeping with [the GDS Way's 'Working with Git' recommendations](https://gds-way.cloudapps.digital/standards/source-code/working-with-git.html#branching-merging-conventions), which say this about handling history in your branch before merging:
> You are encouraged to make liberal use of Git’s [history rewriting features](http://git-scm.com/book/en/Git-Tools-Rewriting-History) while working locally, in order to arrange your commits into appropriate logical chunks that will make sense to your fellow developers. In particular, you may find git rebase --interactive very useful. You are also encouraged to avoid merge commits and use git rebase main instead.

and say this about merging changes into `main`:
> When merging from a feature branch to main (or any other mainline development branch), in particular one that has previously been shared with colleagues, you should use git merge‘s --no-ff option to preserve evidence of your feature branch in the repository history. This advice may be freely ignored for smaller local feature branches for which a fast-forward merge will look like any other routine development work on main.

**Comment by jamie-o-wilkinson** [comment]
Closing as already discussed and decided!
